File: statistics.ml
(* Statistics Calculator in OCaml - Functional Approach *)

(* Calculate mean using List.fold_left *)
let calculate_mean lst =
  if List.length lst = 0 then 0.0
  else
    let sum = List.fold_left (fun acc x -> acc +. float_of_int x) 0.0 lst in
    sum /. float_of_int (List.length lst)

(* Calculate median using List.sort *)
let calculate_median lst =
  if List.length lst = 0 then 0.0
  else
    let sorted = List.sort compare lst in
    let len = List.length sorted in
    if len mod 2 = 0 then
      (* Even number of elements - average of two middle elements *)
      let mid1 = List.nth sorted (len / 2 - 1) in
      let mid2 = List.nth sorted (len / 2) in
      (float_of_int mid1 +. float_of_int mid2) /. 2.0
    else
      (* Odd number of elements - middle element *)
      float_of_int (List.nth sorted (len / 2))

(* Count frequency of each element using List.fold_left *)
let count_frequencies lst =
  List.fold_left
    (fun acc x ->
      if List.mem_assoc x acc then
        List.map (fun (k, v) -> if k = x then (k, v + 1) else (k, v)) acc
      else
        (x, 1) :: acc)
    [] lst

(* Find maximum frequency *)
let find_max_frequency freq_list =
  List.fold_left (fun max_freq (_, freq) -> max max_freq freq) 0 freq_list

(* Find all elements with maximum frequency *)
let find_modes lst =
  if List.length lst = 0 then []
  else
    let freq_list = count_frequencies lst in
    let max_freq = find_max_frequency freq_list in
    List.map fst (List.filter (fun (_, freq) -> freq = max_freq) freq_list)

(* Calculate mode *)
let calculate_mode lst =
  let modes = find_modes lst in
  if List.length modes = List.length (List.sort_uniq compare lst) then
    [] (* No mode if all values occur equally *)
  else
    modes

(* Helper function to print a list *)
let print_list lst =
  print_string "[";
  List.iteri (fun i x ->
    print_int x;
    if i < List.length lst - 1 then print_string ", "
  ) lst;
  print_string "]"

(* Helper function to print modes *)
let print_modes modes =
  if List.length modes = 0 then
    print_string "No mode (all values occur equally)"
  else
    print_list modes

(* Function to read integers from user input *)
let read_integers () =
  print_string "Enter a list of integers separated by spaces: ";
  flush stdout;
  
  let line = read_line () in
  let words = String.split_on_char ' ' line in
  
  let rec parse_integers words acc =
    match words with
    | [] -> List.rev acc
    | word :: rest ->
        try
          let num = int_of_string (String.trim word) in
          parse_integers rest (num :: acc)
        with Failure _ ->
          (* Skip invalid integers *)
          parse_integers rest acc
  in
  
  parse_integers words []

(* Main function *)
let main () =
  print_endline "=== Statistics Calculator (OCaml - Functional) ===";
  print_newline ();
  
  (* Read data from user input *)
  let data = read_integers () in
  
  if List.length data = 0 then (
    print_endline "No valid integers entered.";
    exit 1
  );
  
  print_string "Input data: ";
  print_list data;
  print_newline ();
  print_newline ();
  
  (* Calculate and display statistics *)
  let mean = calculate_mean data in
  Printf.printf "Mean: %.2f\n" mean;
  
  let median = calculate_median data in
  Printf.printf "Median: %.2f\n" median;
  
  let modes = calculate_mode data in
  print_string "Mode: ";
  print_modes modes;
  print_newline ()

(* Run the main function *)
let () = main () 

File: statistics.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Function to calculate mean of integers
double calculate_mean(int arr[], int size) {
    if (size == 0) return 0.0;
    
    double sum = 0.0;
    for (int i = 0; i < size; i++) {
        sum += arr[i];
    }
    return sum / size;
}

// Function to compare integers for qsort
int compare_ints(const void* a, const void* b) {
    return (*(int*)a - *(int*)b);
}

// Function to calculate median of integers
double calculate_median(int arr[], int size) {
    if (size == 0) return 0.0;
    
    // Create a copy of the array to sort
    int* sorted_arr = malloc(size * sizeof(int));
    if (sorted_arr == NULL) {
        fprintf(stderr, "Memory allocation failed\n");
        return 0.0;
    }
    
    memcpy(sorted_arr, arr, size * sizeof(int));
    qsort(sorted_arr, size, sizeof(int), compare_ints);
    
    double median;
    if (size % 2 == 0) {
        // Even number of elements - average of two middle elements
        median = (sorted_arr[size/2 - 1] + sorted_arr[size/2]) / 2.0;
    } else {
        // Odd number of elements - middle element
        median = sorted_arr[size/2];
    }
    
    free(sorted_arr);
    return median;
}

// Function to calculate mode of integers
void calculate_mode(int arr[], int size, int* modes, int* mode_count) {
    if (size == 0) {
        *mode_count = 0;
        return;
    }
    
    // Find the range of values
    int min_val = arr[0], max_val = arr[0];
    for (int i = 1; i < size; i++) {
        if (arr[i] < min_val) min_val = arr[i];
        if (arr[i] > max_val) max_val = arr[i];
    }
    
    int range = max_val - min_val + 1;
    int* frequency = calloc(range, sizeof(int));
    if (frequency == NULL) {
        fprintf(stderr, "Memory allocation failed\n");
        *mode_count = 0;
        return;
    }
    
    // Count frequency of each value
    for (int i = 0; i < size; i++) {
        frequency[arr[i] - min_val]++;
    }
    
    // Find maximum frequency
    int max_freq = 0;
    for (int i = 0; i < range; i++) {
        if (frequency[i] > max_freq) {
            max_freq = frequency[i];
        }
    }
    
    // Find all values with maximum frequency
    *mode_count = 0;
    for (int i = 0; i < range; i++) {
        if (frequency[i] == max_freq) {
            modes[*mode_count] = i + min_val;
            (*mode_count)++;
        }
    }
    
    // If all values occur equally, return no mode
    if (*mode_count == range) {
        *mode_count = 0;
    }
    
    free(frequency);
}

// Function to print array
void print_array(int arr[], int size) {
    printf("[");
    for (int i = 0; i < size; i++) {
        printf("%d", arr[i]);
        if (i < size - 1) printf(", ");
    }
    printf("]\n");
}

// Function to print modes
void print_modes(int modes[], int mode_count) {
    if (mode_count == 0) {
        printf("No mode (all values occur equally)\n");
        return;
    }
    
    printf("[");
    for (int i = 0; i < mode_count; i++) {
        printf("%d", modes[i]);
        if (i < mode_count - 1) printf(", ");
    }
    printf("]\n");
}

// Function to read integers from user input
int read_integers(int arr[], int max_size) {
    printf("Enter a list of integers separated by spaces: ");
    
    char line[1000];
    if (fgets(line, sizeof(line), stdin) == NULL) {
        return 0;
    }
    
    int count = 0;
    char* token = strtok(line, " \t\n");
    
    while (token != NULL && count < max_size) {
        arr[count] = atoi(token);
        count++;
        token = strtok(NULL, " \t\n");
    }
    
    return count;
}

int main() {
    printf("=== Statistics Calculator (C - Procedural) ===\n\n");
    
    int data[1000]; // Maximum 1000 integers
    int size = read_integers(data, 1000);
    
    if (size == 0) {
        printf("No valid integers entered.\n");
        return 1;
    }
    
    printf("Input data: ");
    print_array(data, size);
    printf("\n");
    
    // Calculate and display statistics
    double mean = calculate_mean(data, size);
    printf("Mean: %.2f\n", mean);
    
    double median = calculate_median(data, size);
    printf("Median: %.2f\n", median);
    
    int modes[100]; // Assuming max 100 different modes
    int mode_count;
    calculate_mode(data, size, modes, &mode_count);
    printf("Mode: ");
    print_modes(modes, mode_count);
    
    return 0;
} 